grovedb/src/batch/multi_insert_cost_tests.rs

```diff
@@ -0,0 +1,270 @@
+                b"key3",
+                Element::empty_tree(),
+                None,
+                Some(&tx),
+            )
+            .cost;
+        let non_batch_cost_4 = db
+            .insert(
+                vec![b"key1".as_slice(), b"key3".as_slice()],
+                b"key4",
+                Element::new_reference(UpstreamFromElementHeightReference(
+                    1,
+                    vec![b"key2".to_vec()],
+                )),
+                None,
+                Some(&tx),
+            )
+            .cost;
+        let non_batch_cost = non_batch_cost_1
+            .add(non_batch_cost_2)
+            .add(non_batch_cost_3)
+            .add(non_batch_cost_4);
+        tx.rollback().expect("expected to rollback");
+        let ops = vec![
+            GroveDbOp::insert_run_op(vec![], b"key1".to_vec(), Element::empty_tree()),
+            GroveDbOp::insert_run_op(
+                vec![b"key1".to_vec()],
+                b"key2".to_vec(),
+                Element::new_item_with_flags(b"pizza".to_vec(), Some([0, 1].to_vec())),
+            ),
+            GroveDbOp::insert_run_op(
+                vec![b"key1".to_vec()],
+                b"key3".to_vec(),
+                Element::empty_tree(),
+            ),
+            GroveDbOp::insert_run_op(
+                vec![b"key1".to_vec(), b"key3".to_vec()],
+                b"key4".to_vec(),
+                Element::new_reference(UpstreamFromElementHeightReference(
+                    1,
+                    vec![b"key2".to_vec()],
+                )),
+            ),
+        ];
+        let cost = db
+            .apply_batch(ops, None, Some(&tx))
+            .cost_as_result()
+            .expect("expected to apply batch");
+        assert_eq!(
+            non_batch_cost.storage_cost.added_bytes,
+            cost.storage_cost.added_bytes
+        );
+        assert_eq!(non_batch_cost.storage_cost.removed_bytes, NoStorageRemoval);
+        assert_eq!(cost.storage_cost.removed_bytes, NoStorageRemoval);
+    }
+
+    #[test]
+    fn test_batch_root_two_insert_tree_cost_same_level() {
+        let db = make_empty_grovedb();
+        let tx = db.start_transaction();
+
+        let ops = vec![
+            GroveDbOp::insert_run_op(vec![], b"key1".to_vec(), Element::empty_tree()),
+            GroveDbOp::insert_run_op(vec![], b"key2".to_vec(), Element::empty_tree()),
+        ];
+        let cost_result = db.apply_batch(ops, None, Some(&tx));
+        cost_result.value.expect("expected to execute batch");
+        let cost = cost_result.cost;
+        // Explanation for 214 storage_written_bytes
+
+        // Key -> 37 bytes
+        // 32 bytes for the key prefix
+        // 4 bytes for the key
+        // 1 byte for key_size (required space for 36)
+
+        // Value -> 37
+        //   1 for the flag option (but no flags)
+        //   1 for the enum type
+        //   1 for empty tree value
+        // 32 for node hash
+        // 0 for value hash
```

Comment:
Sure, the "value hash" is either the hash of the value for an item, the hash of the referenced item for a reference or the root hash for a tree.
the hash(hash(value) + value_hash) = elem_value_hash
hash(key + elem_value_hash) = kv_hash
hash(kv_hash + left + right) = node_hash

We keep the value_hash, because we need it to be able to prove various things such as references.

Reply:
hash(value) != value_hash for references but hash(value) == value_hash for ordinary value?

Reply:
@iammadab please confirm.

Reply:
Yeah, hash(value) != value_hash for references and trees but hash(value) == value_hash for Items.